sample/rules_test.go

```diff
@@ -2152,3 +2157,426 @@ func TestRulesWithDeterministicSampler(t *testing.T) {
+						Data: map[string]interface{}{
+							"http.status_code": "200",
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"test":  "nope",
+							"test1": 1,
+							"test2": 2.2,
+						},
+					},
+				},
+			},
+			ExpectedKeep: true,
+			ExpectedRate: 1,
+			ExpectedName: "no rule matched",
+		},
+		{
+			Rules: &config.RulesBasedSamplerConfig{
+				Rules: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "condition with multiple fields",
+						SampleRate: 10,
+						Conditions: []*config.RulesBasedSamplerCondition{
+							{
+								Fields:   []string{"field1", "field2", "root.field3"},
+								Operator: config.Contains,
+								Value:    "foo",
+								Datatype: "string",
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					TraceID: "123testABC", // I am root.
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"field1": "foo",
+							"field2": "foo",
+							"field3": "foo",
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"http.status_code": "200",
+						},
+					},
+				},
+			},
+			ExpectedKeep: false,
+			ExpectedRate: 10,
+			ExpectedName: "condition with multiple fields",
+		},
+		{
+			Rules: &config.RulesBasedSamplerConfig{
+				Rules: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "condition with multiple fields, does not match",
+						SampleRate: 10,
+						Conditions: []*config.RulesBasedSamplerCondition{
+							{
+								Fields:   []string{"field1", "field2", "root.field3"},
+								Operator: config.EQ,
+								Value:    1.0,
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					TraceID: "123testABC", // I am root.
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"field3": 1.5,
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"field1": 1.0,
+						},
+					},
+				},
+			},
+			ExpectedKeep: true,
+			ExpectedRate: 1,
+			ExpectedName: "no rule matched",
+		},
+	}
+
+	for _, d := range data {
+		t.Run(d.Rules.Rules[0].Name, func(t *testing.T) {
+			sampler := &RulesBasedSampler{
+				Config:  d.Rules,
+				Logger:  &logger.NullLogger{},
+				Metrics: &metrics.NullMetrics{},
+			}
+
+			sampler.Start()
+
+			trace := &types.Trace{}
+			for _, span := range d.Spans {
+				trace.AddSpan(span)
+				// We declare which span is the root span in the test cases by setting the traceID.
+				if span.TraceID != "" {
+					require.Nil(t, trace.RootSpan, "Only set the trace ID on one span in a test case to designate which is the root. This test case appears to have multiple spans with trace IDs set.")
+					trace.RootSpan = span
+				}
+			}
+
+			spans := trace.GetSpans()
+			assert.Len(t, spans, len(d.Spans), "should have the same number of spans as input")
+
+			rate, _, reason, key := sampler.GetSampleRate(trace)
```

Comment:
It does not make sense to test on `ExpectedKeep` here because it makes the test flakey. It we are sampling values at 1 in 10 then `ExpectedKeep` will fail on every tenth test. We do test `keep` in other tests.